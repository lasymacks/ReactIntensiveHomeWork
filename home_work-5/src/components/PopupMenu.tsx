import  React, { useCallback, useMemo, useState } from 'react';
import { useEffect } from 'react';
import ReactDOM from 'react-dom';
 
interface IPropsPopup {
   isOpen: boolean,
   id: number,
   setIsOpen: any,
   children: any
}

// const modalRootElement = document.querySelector("#modal");

// Привет, Макс! Я очень долго боролся с CreatePortal и сдался(на второй день), видимо уже так запутался,
// что распутаться не представляется возможным)) У меня вопрос, ответь пожалуйста когда 
// когда будешь проверять. Почему мой портал не вызывается с обновлением стейта?
// Я пробовал много разных вариантов, но ни один не сработал корректно. Я вроде бы все сделал
// верно, а перерендер не вызывал. В DevTools появлялся компонент, но был пустой.
// Думал обойдусь прост обновлением стейта, но ошибался. Заранее спасибо!
// P.S. Я специально оставил комменты кода от портала, чтобы ты смог посмотреть где я накосячил.

const PopupItem = (props: IPropsPopup) => {

    const isOpen = props.isOpen;
	const element = useMemo(() => document.createElement("div"), []);

	// useEffect(() => {
	// 	if (isOpen) {      
	// 		modalRootElement?.appendChild(element);
	// 	}
	// }, [isOpen]);

    // useEffect(() =>  {
    //     return () => {
    //         modalRootElement?.removeChild(element);
    //     };
    // }, [isOpen])

    const closeHandler = () => {
        props.setIsOpen(() => false)
    }

	if (isOpen) {
        // ReactDOM.createPortal
		return (
            <div className='popup-backgroung' onClick={closeHandler}>
                    {props.children}
            </div>
			// ,
			// element
		);
	}

	return null;
    
}

export default PopupItem;
