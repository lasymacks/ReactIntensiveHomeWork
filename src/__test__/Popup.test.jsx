import PopupMenu from "../components/PopupMenu";
import React from "react";

describe('Popup', () => {
    let wrapper;
    const onClick = jest.fn();
    const props = {
        isOpen: true
    }

    beforeEach(() => {
        wrapper = mount(<PopupMenu {...props} />);
    });

    afterEach(() => {
        wrapper.unmount();
    });

    it('component should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
        console.log(wrapper.debug());
    });

    it('component should call a function onBlur with we lose focus', () => {
        const div = shallow(<div data-test-id='popup' className='popup-backgroung' onClick={onClick} />);
        div.find('[data-test-id="popup"]').simulate('click');
        expect(onClick.mock.calls.length).toEqual(1);
    })

});