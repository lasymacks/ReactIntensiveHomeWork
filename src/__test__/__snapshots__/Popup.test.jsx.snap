// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup component should render correctly 1`] = `ReactWrapper {}`;
